File: LeetCode/Explore/TopInterview/Amazon/ArrayString/238_product_of_array_except_self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
      int mul = 1;
      vector<int> dp(nums.size(), 1);
      
      for (int i = 0; i < nums.size(); i++) {
        dp[i] *= mul;
        mul *= nums[i];
      }
      
      mul = 1;
      for (int i = nums.size() - 1; i >= 0; i--) {
        dp[i] *= mul;
        mul *= nums[i];
      }
      
      return dp;
    }
};

File: LeetCode/Explore/TopInterview/Amazon/ArrayString/238_product_of_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        mul = 1
        dp = [1] * len(nums)
        
        for i, num in enumerate(nums):
          dp[i] *= mul
          mul *= num
          
        mul = 1  
        for i in range(len(nums) - 1, -1, -1):
          dp[i] *= mul
          mul *= nums[i]
          
        return dp
